File: src-tauri/src/agents/echo.rs
use async_trait::async_trait;

use super::backend::{AgentBackend, BackendConfig, BackendMessage, BackendResponse, MessageRole};

pub struct EchoBackend;

#[async_trait]
impl AgentBackend for EchoBackend {
    fn id(&self) -> &str {
        "echo"
    }

    fn display_name(&self) -> &str {
        "Echo (NPC)"
    }

    async fn respond(
        &self,
        config: &BackendConfig,
        messages: &[BackendMessage],
    ) -> Result<BackendResponse, String> {
        let last_user_msg = messages
            .iter()
            .rev()
            .find(|m| m.role == MessageRole::User)
            .map(|m| m.content.as_str())
            .unwrap_or("");

        let content = match &config.system_prompt {
            Some(prompt) if !prompt.is_empty() => {
                format!("{} Echo: {}", prompt, last_user_msg)
            }
            _ => format!("Echo: {}", last_user_msg),
        };

        Ok(BackendResponse {
            content,
            needs_attention: false,
        })
    }

    async fn is_available(&self) -> bool {
        true
    }

    fn as_any(&self) -> &dyn std::any::Any {
        self
    }
}

File: .worktree/vscode-extension/vscode-extension/src/agents/echo.ts
import type {
  AgentBackend,
  BackendConfig,
  BackendMessage,
  BackendResponse,
  ModelOption,
  AgentOption,
} from "./backend.js";

export class EchoBackend implements AgentBackend {
  id(): string {
    return "echo";
  }

  displayName(): string {
    return "Echo (NPC)";
  }

  async respond(
    _agentId: string,
    config: BackendConfig,
    messages: BackendMessage[],
  ): Promise<BackendResponse> {
    const lastUserMsg = [...messages]
      .reverse()
      .find((m) => m.role === "user");
    const echo = lastUserMsg ? lastUserMsg.content : "";

    const content =
      config.systemPrompt && config.systemPrompt.length > 0
        ? `${config.systemPrompt} Echo: ${echo}`
        : `Echo: ${echo}`;

    return { content, needsAttention: false };
  }

  async destroyChatSession(_agentId: string): Promise<void> {}

  async isAvailable(): Promise<boolean> {
    return true;
  }

  async listModels(): Promise<ModelOption[]> {
    return [];
  }

  async listAgents(_cwd?: string): Promise<AgentOption[]> {
    return [];
  }

  async setApiKey(_key: string): Promise<void> {}

  credentialKey(): string | undefined {
    return undefined;
  }
}
